test/test_translational_smoothing.py

```diff
@@ -116,50 +106,60 @@ def generate_test_description():
 
 class TestCommandProfile(unittest.TestCase):
 
+    @classmethod
+    def setUpClass(cls):
+        rclpy.init()
+
+    @classmethod
+    def tearDownClass(cls):
+        rclpy.shutdown()
+
+    def setUp(self):
+        self.node = rclpy.create_node('test_velocity_smoother')
+
+    def tearDown(self):
+        self.node.destroy_node()
+
     def test_subscribe_vel_topics(self, launch_service, commands, proc_output):
-        launch_context = launch_service.context
-        node = launch_context.locals.launch_ros_node
         input_velocities = []
         input_timestamps = []
         smoothed_velocities = []
         smoothed_timestamps = []
 
         def received_input_data(msg):
-            # node.get_logger().warn("received input data {}".format(msg))
             input_velocities.append(msg.linear.x)
             input_timestamps.append(time.monotonic())
 
         def received_smoothed_data(msg):
-            # node.get_logger().warn("received smoothed data {}".format(msg))
             smoothed_velocities.append(msg.linear.x)
             smoothed_timestamps.append(time.monotonic())
 
-        input_subscriber = node.create_subscription(
+        input_subscriber = self.node.create_subscription(
             geometry_msgs.msg.Twist,
             '/commands/cmd_vel',
             received_input_data,
             10,
         )
-        smoothed_subscriber = node.create_subscription(
+        smoothed_subscriber = self.node.create_subscription(
             geometry_msgs.msg.Twist,
             '/cmd_vel',
             received_smoothed_data,
             10,
         )
         try:
-            node.get_logger().info("Waiting for PROFILE_SENT")
-            proc_output.assertWaitFor(expected_output="PROFILE_SENT", process=commands, timeout=60)
+            done = False
+            while rclpy.ok() and not done:
+                rclpy.spin_once(self.node, timeout_sec=0.1)
+                try:
+                    assertInStdout(proc_output, "PROFILE_SENT", commands)
```

Comment:
Isn't this brittle? There's no guarantee that you've actually got data from the subscribers at this point. Probably better off waiting for that data with a timeout for failure?

Reply:
The way that this works now is that we call `spin_once`, followed by `assertInStdout`.  We do this so that we ensure that we are calling callbacks and otherwise letting the executor do its work.  That's why we can't block here with a timeout.  In this system, `assertInStdout` can (and does) throw asserts often.  But we catch them down below and ignore them.  Once we see "PROFILE_SENT", we then go on to do the rest of the testing.  We are prevented from running forever by the overall timeout set on the test for `ctest` in the CMakeLists.txt; I arbitrarily set that to 30 seconds, but we can lower that bounds if needed.

Reply:
What guarantees that the callback actually fired in the same `spin_once` as the publication? Is it because you're working on the assumption that it's all inter-process?

I would have just conditioned that loop on `input_velocities` being non-empty. Seems a bit less obscure than depending on stdout.

Reply:
> What guarantees that the callback actually fired in the same `spin_once` as the publication? 

Hm, I think we are talking about different things here.

The way that this test is setup is that there are two external processes that are being created: `translational_command_profile.py`, and an instance of `velocity_smoother`.  Each of those is launched, and each has their own rclpy and executor within them.  `translational_command_profile.py` creates two publishers, one subscriber, and one timer.  Every 10th of a second, it publishes some data from a fixed list into the `/commands/cmd_vel`.  `velocity_smoother` subscribes to that `/commands/cmd_vel`, does its smoothing, and then publishes the new data on `/cmd_vel`.  `translational_command_profile.py` gets a callback on `/cmd_vel` data, and then publishes some of that data as a fake `/odometry` topic.  `velocity_smoother` then uses that fake `/odometry` to do its internal calculations.

Note that *none* of the above involves `test_translational_smoothing.py` at all, except for the fact that the test kicks off the two processes.  Once it kicks them off, in some sense it could just wait for them to finish and exit.  However, to get a better idea of what is going on, `test_translational_smoothing.py` also subscribes to some of these topics to introspect what is going on (in particular `/commands/cmd_vel` and `/cmd_vel`).  Thus, it is also only a single-threaded executor in its own (third) process, and thus it has to call `rclpy.spin` to ensure that the callbacks for its own subscriptions are processed.  Once it has determined that the process has finished (via the 'PROFILE_SENT' stdout print from `test_translational_command_profile.py`), it can then check that it got everything it expected.

Given how this is setup, I think we need all of these pieces.  One thing we *could* do is change 'PROFILE_SENT' from a print to a subscription; that would avoid all of the `try..except` shenanigans.  But it is extra complexity, so I'm not sure it is worth it.

Also, FYI, this is modeled after https://github.com/ros2/launch_ros/blob/master/launch_testing_ros/test/examples/talker_listener_launch_test.py .

Does this all make sense to you?  Let me know if you have further thoughts on how to proceed here.

Reply:
Forgetting the brain lapse of the intra-process comment, I was wondering if something like:

```
self.assertEqual(90, len(input_velocities))
```

would have broken this test. I'd assumed you had printed the `PROFILE_SENT` immediately after publishing the last data point, at which point it would have been a race between the stdout parsing and the subscription to make that assert work.

As it is, it's still a race, but you give the `PROFILE_SENT` the chance to win because you print that 0.1s after the last publication. I guess it's more efficient, because any way you write this test it's going to need some hard coded time delta within which you're expecting operations to finish and this delta is encoded at the latest possible point, but ... having test configuration/logic outside of the test (i.e. over in `translational_command.py`) makes this test harder to understand & debug.

Reply:
I added a comment in-code to help anyone who visits this code in the future.

Reply:
> I'd still argue that you don't lose anything by conditioning on the length of the input velocities to determine when to break out into assertion checking, except maybe efficiency in failing faster if connections aren't working properly, but you gain simpler test logic.

Ah yes, I see what you mean now.  So the most basic reason not to do this is so that you can arbitrarily change how many things are in `translational_command_profile.py` without affecting how the loop works.  I think if I were going to revamp this test again, I would actually move all of `translational_command_profile.py` into `test_translational_smoothing.py`, and reduce the number of processes/executors to 2.  I think that would make this somewhat easier to understand, though it is still convoluted because of the back and forth between all of the processes.